// @sanity/cli/lib/commands/projects/listProjectsCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const {
  size,
  sortBy
} = require('lodash');

const headings = ['id', 'members', 'name', 'url', 'created'];
const helpText = `
Options
  --sort <field> Sort output by specified column
  --order <asc/desc> Sort output ascending/descending

Examples
  # List projects
  sanity projects list

  # List projects sorted by member count, ascending
  sanity projects list --sort=members --order=asc
`;
const defaultFlags = {
  sort: 'created',
  order: 'desc'
};
var _default = {
  name: 'list',
  group: 'projects',
  signature: '',
  helpText,
  description: 'Lists projects connected to your user',
  action: async (args, context) => {
    const {
      apiClient,
      output,
      chalk
    } = context;

    const flags = _objectSpread({}, defaultFlags, {}, args.extOptions);

    const client = apiClient({
      requireUser: true,
      requireProject: false
    });
    const projects = await client.request({
      method: 'GET',
      uri: '/projects'
    });
    const ordered = sortBy(projects.map(({
      displayName,
      id,
      members = [],
      studioHost = '',
      createdAt
    }) => {
      const studio = studioHost ? `https://${studioHost}.sanity.studio` : 'Not deployed';
      return [id, members.length, displayName, studio, createdAt];
    }), [headings.indexOf(flags.sort)]);
    const rows = flags.order === 'asc' ? ordered : ordered.reverse();
    const maxWidths = rows.reduce((max, row) => row.map((current, index) => Math.max(size(current), max[index])), headings.map(str => size(str)));

    const printRow = row => row.map((col, i) => `${col}`.padEnd(maxWidths[i])).join('   ');

    output.print(chalk.cyan(printRow(headings)));
    rows.forEach(row => output.print(printRow(row)));
  }
};
exports.default = _default;
