// @quentin-sommer/react-useragent/lib/Provider.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uaParserJs = require('ua-parser-js');

var _uaParserJs2 = _interopRequireDefault(_uaParserJs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UAProvider = function (_React$Component) {
  _inherits(UAProvider, _React$Component);

  function UAProvider(props, context) {
    _classCallCheck(this, UAProvider);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props, context));

    var uaParser = new _uaParserJs2.default();
    var uaResults = {
      android: false,
      ios: false,
      mobile: false,
      tablet: false,
      windows: false,
      mac: false,
      linux: false,
      computer: false,
      firefox: false,
      chrome: false,
      edge: false,
      safari: false
    };

    uaParser.setUA(props.ua);
    uaResults.android = uaParser.getOS().name === 'Android';
    uaResults.ios = uaParser.getOS().name === 'iOS';
    uaResults.mobile = uaParser.getDevice().type === 'mobile';

    uaResults.tablet = uaParser.getDevice().type === 'tablet';

    uaResults.windows = uaParser.getOS().name === 'Windows';
    uaResults.mac = uaParser.getOS().name === 'Mac OS';
    uaResults.linux = uaParser.getOS().name === 'Linux';
    uaResults.computer = uaResults.windows || uaResults.mac || uaResults.linux || uaParser.getDevice().type === undefined;

    uaResults.firefox = uaParser.getBrowser().name === 'Firefox';
    uaResults.chrome = uaParser.getBrowser().name === 'Chrome';
    uaResults.edge = uaParser.getBrowser().name === 'Edge';
    uaResults.safari = uaParser.getBrowser().name === 'Safari';

    _this.uaParser = uaParser;
    _this.uaResults = uaResults;
    return _this;
  }

  UAProvider.prototype.getChildContext = function getChildContext() {
    return {
      ua: {
        parser: this.uaParser,
        uaResults: this.uaResults
      }
    };
  };

  UAProvider.prototype.render = function render() {
    return _react2.default.Children.only(this.props.children);
  };

  return UAProvider;
}(_react2.default.Component);

if (process.env.NODE_ENV !== 'production') {
  UAProvider.propTypes = process.env.NODE_ENV !== "production" ? {
    ua: _propTypes2.default.string.isRequired,
    children: _propTypes2.default.element.isRequired
  } : {};
}

UAProvider.childContextTypes = {
  ua: _propTypes2.default.object.isRequired
};

exports.default = UAProvider;
module.exports = exports['default'];
