// @sanity/base/lib/components/NoJavascript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NoJavascript;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NoJsStyles = `
.sanity-app-no-js__root {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background: #fff;
}

.sanity-app-no-js__content {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  font-family: helvetica, arial, sans-serif;
}
`;

function NoJavascript() {
  return _react.default.createElement("noscript", null, _react.default.createElement("div", {
    className: "sanity-app-no-js__root"
  }, _react.default.createElement("div", {
    className: "sanity-app-no-js__content"
  }, _react.default.createElement("style", {
    type: "text/css"
  }, NoJsStyles), _react.default.createElement("h1", null, "JavaScript disabled"), _react.default.createElement("p", null, "Please ", _react.default.createElement("a", {
    href: "https://www.enable-javascript.com/"
  }, "enable JavaScript"), " in your browser and reload the page to proceed."))));
}
